#162948907 generate and return jwt tokens

danuluma opened the pull request:
#### What does this PR do?
Users should receive a JWT on successful Registration 

#### Description of Task(s) to be completed?
return a JWT to a user on successful registration
generate and return a JWT when a user logs in

#### How should this be manually tested?
* Pull the branch `git fetch origin ft-return-token-162948907:ft-return-token-162948907`
* Checkout to the branch `git checkout ft-return-token-162948907`
* Install dependencies `pip3 install -r requirements.txt`
* Run the server `python3 manage.py runserver`
* Use postman to send a `POST` request to `http://127.0.0.1:8000/api/users/`
* The data should be in the format of:
```
{   "user":{
    "username": "dann",
    "email": "dann@dan.com",
    "password": "county001"
    }
 }
```
* Remember to select `application/json` as the data type
* The response should be a json object containing the user's details as well as the jwt token
```
{
    "user": {
        "email": "dann@dan.com",
        "username": "dann",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoxNiwidXNlcm5hbWUiOiJkYW5uQGRhbi5jb20iLCJleHAiOjE1NDc1NzgyMDgsImVtYWlsIjoiZGFubkBkYW4uY29tIn0.n8mW-ncEXc37EwALIlQgF6CVyMtdjKWuowy6SYjamPA"
    }
}
```


#### Any background context you want to provide?
* This project uses the latest version of python3. Please download it from [here](https://www.python.org/downloads/)
* You may as well use pip instead of pip3 if you have either only python3 installed on your computer or you have it set as the default version of python
* Postman can be downloaded from [here](https://www.getpostman.com/downloads/)
* This token needs to be sent in the headers of subsequent requests to protected endpoints
* The header prefix expected is `Bearer`. On postman, this would hence be sent as a Bearer Token
* I also downgraded django-rest-swagger from v2.2.0 to v2.1.2 due to a bug where the authorization headers aren't passed in the request. It is well described [here](https://github.com/marcgibbons/django-rest-swagger/issues/762).

#### What are the relevant pivotal tracker stories?
[#162948907](https://www.pivotaltracker.com/story/show/162948907)

Commit:
feature(validation): Recieve descritive validation error message

- Users should recieve a descriptive error message on registrattion and login

[Delivers #162948908]

Ruiru11 reviewed (commented):
We agreed we should have one commit

danuluma commented:
I'm fixing that

danuluma committed:
feature(jwt): generate and return jwt tokens

- Generate and return a jwt token when a user registers
- Create and return a jwt token on login

[Delivers #162948907]

tesh254 commented on authors/apps/authentication/backends.py:
```diff
@@ -1,10 +1,31 @@
 """Configure JWT Here"""
 
+
+class JWTokens(object):
```
This class should have a docstring as a PEP8 standard

danuluma replied:
Thanks,,,,correcting that

tesh254 commented on authors/apps/authentication/models.py:
```diff
@@ -4,117 +4,122 @@
 from django.conf import settings
 from django.contrib.auth.models import (
-    AbstractBaseUser, BaseUserManager, PermissionsMixin
+		AbstractBaseUser, BaseUserManager, PermissionsMixin
```
Also this space right please fix them

danuluma replied:
OK

tesh254 commented on authors/apps/authentication/models.py:
```diff
@@ -4,117 +4,122 @@
+    return user
 
 
 class User(AbstractBaseUser, PermissionsMixin):
```
Lacks a docstring to describe the class

danuluma replied:
Please do check again

Ruiru11 commented on authors/apps/authentication/serializers.py:
```diff
@@ -23,7 +52,48 @@ class Meta:
+
+    def validate_username(self, value):
+        #get initial state of data as an object and pick the value of username
+        data = self.get_initial()
```
we agreed we are getting rid of this

Ruiru11 commented on authors/apps/authentication/serializers.py:
```diff
@@ -34,7 +104,7 @@ class LoginSerializer(serializers.Serializer):
+    token = serializers.CharField(max_length=128, read_only=True)
 
     def validate(self, data):
         # The `validate` method is where we make sure that the current
```
This comment is incomplete

danuluma replied:
Not quite, please re-check

frankip commented on authors/apps/authentication/models.py:
```diff
@@ -6,115 +6,120 @@
 )
-from django.db import models 
+from django.db import models
+# from rest_framework_jwt.settings import api_settings
```
@danuluma  Remove commented out code

danuluma replied:
I've done that, thanks

frankip reviewed (approved):
LGTM
